File: javascript/mlp/scripts/Mlp.js
class Mlp {

    constructor(numberOfInputs, neuronsPerLayer, learningRate) {
        this.neuronLayers = [];
        for (var i = 0; i < neuronsPerLayer.length; i++) {
            var neuronLayer = this.initializeNeuronLayer(numberOfInputs, neuronsPerLayer[i], i);
            this.neuronLayers.push(neuronLayer);
        }
        this.learningRate = learningRate;
        this.debug = true;
    }

    initializeNeuronLayer(numberOfInputs, numberOfNeurons, numberOflayer) {
        var ret = [];
        for(var i = 0; i < numberOfNeurons; i++) {
            var neuron;
            if(numberOflayer === 0) {
                neuron = new Neuron(numberOfInputs);
            } else {
                var middleLayerNumberOfInputs = this.neuronLayers[numberOflayer - 1].length;
                neuron = new Neuron(middleLayerNumberOfInputs);
            }
            ret.push(neuron);
        }
        return ret;
    }

    calculateOutputs(inputs) {
        for(var i = 0; i < this.neuronLayers.length; i++) {
            for(var j = 0; j < this.neuronLayers[i].length; j++) {
                if(i === 0) {
                    this.neuronLayers[i][j].calculateOutput(inputs);
                } else {
                    var middleLayerInputs = this.getLayerOutputs(this.neuronLayers[i - 1]);
                    this.neuronLayers[i][j].calculateOutput(middleLayerInputs);
                }
            }
        }
        return this.getLayerOutputs(this.neuronLayers[this.neuronLayers.length - 1]);
    }

    getLayerOutputs(neuronLayer) {
        var ret = [];
        for(var i = 0; i < neuronLayer.length; i++) {
            ret.push(neuronLayer[i].getOutput());
        }
        return ret;
    }

    train(trainingElement) {
        this.calculateOutputs(trainingElement.inputs);
        // train output layer
        var outputLayer = this.neuronLayers.length - 1;
        this.trainOutputLayer(trainingElement.desiredOutputs, this.neuronLayers[outputLayer], this.neuronLayers[outputLayer - 1]);
        // train inner layers
        for(var i = this.neuronLayers.length - 2; i > 0; i--) {
            this.trainInnerLayer(this.neuronLayers[i], this.neuronLayers[i - 1], this.neuronLayers[i + 1]);
        }
        // train input layer
        this.trainInputLayer(this.neuronLayers[0], trainingElement.inputs, this.neuronLayers[1]);
    }

    trainOutputLayer(desiredOutputs, currentLayer, previousLayer) {
        for(var i = 0; i < currentLayer.length; i++) {
            var currentNeuron = currentLayer[i];
            for(var j = 0; j < currentNeuron.getWeights().length; j++) {
                var jOutput = currentNeuron.getOutput();
                var jTargetOutput = desiredOutputs[i];
                var iOutput = previousLayer[j].getOutput();
                var jDeltaOutput = (jOutput - jTargetOutput) * jOutput * (1 - jOutput);
                var ijDeltaWeight = -this.learningRate  * iOutput * jDeltaOutput;
                currentNeuron.setDeltaOutput(jDeltaOutput);
                currentNeuron.adjustWeight(ijDeltaWeight, j);
            }
            // Train bias:
            var jOutput = currentNeuron.getOutput();
            var jTargetOutput = desiredOutputs[i];
            var iOutput = 1;
            var jDeltaOutput = (jOutput - jTargetOutput) * jOutput * (1 - jOutput);
            var ijDeltaWeight = -this.learningRate  * iOutput * jDeltaOutput;
            currentNeuron.adjustBias(ijDeltaWeight);

        }
    }

    trainInnerLayer(currentLayer, previousLayer, nextLayer) {
        for(var i = 0; i < currentLayer.length; i++) {
            var currentNeuron = currentLayer[i];
            for(var j = 0; j < currentNeuron.getWeights().length; j++) {
                var jOutput = currentNeuron.getOutput();
                var iOutput = previousLayer[j].getOutput();
                var summation = 0;
                for(var l = 0; l  < nextLayer.length; l++) {
                    summation += nextLayer[l].getDeltaOutput() * nextLayer[l].getWeight(i);
                }
                var jDeltaOutput = summation * jOutput * (1 - jOutput);
                var ijDeltaWeight = -this.learningRate * iOutput * jDeltaOutput;
                currentNeuron.setDeltaOutput(jDeltaOutput);
                currentNeuron.adjustWeight(ijDeltaWeight, j);
            }
            // Train bias:
            var jOutput = currentNeuron.getOutput();
            var iOutput = 1;
            var summation = 0;
            for(var l = 0; l  < nextLayer.length; l++) {
                summation += nextLayer[l].getDeltaOutput() * nextLayer[l].getWeight(i);
            }
            var jDeltaOutput = summation * jOutput * (1 - jOutput);
            var ijDeltaWeight = -this.learningRate  * iOutput * jDeltaOutput;
            currentNeuron.adjustBias(ijDeltaWeight);
        }
    }

    trainInputLayer(currentLayer, inputs, nextLayer) {
        for(var i = 0; i < currentLayer.length; i++) {
            var currentNeuron = currentLayer[i];
            for(var j = 0; j < currentNeuron.getWeights().length; j++) {
                var jOutput = currentNeuron.getOutput();
                var iOutput = inputs[j];
                var summation = 0;
                for(var l = 0; l  < nextLayer.length; l++) {
                    summation += nextLayer[l].getDeltaOutput() * nextLayer[l].getWeight(i);
                }
                var jDeltaOutput = summation * jOutput * (1 - jOutput);
                var ijDeltaWeight = -this.learningRate  * iOutput * jDeltaOutput;
                currentNeuron.adjustWeight(ijDeltaWeight, j);
            }
            // Train bias:
            var jOutput = currentNeuron.getOutput();
            var iOutput = 1;
            var summation = 0;
            for(var l = 0; l  < nextLayer.length; l++) {
                summation += nextLayer[l].getDeltaOutput() * nextLayer[l].getWeight(i);
            }
            var jDeltaOutput = summation * jOutput * (1 - jOutput);
            var ijDeltaWeight = -this.learningRate  * iOutput * jDeltaOutput;
            currentNeuron.adjustBias(ijDeltaWeight);
        }
    }

    print() {
        for(var i = 0; i < this.neuronLayers.length; i++) {
            for(var j = 0; j < this.neuronLayers[i].length; j++) {
                console.log("[" + i + "][" + j + "]");
                this.neuronLayers[i][j].print();
            }
        }
    }

}

File: java/main/mlp/src/main/java/com/yzumis/ia/mlp/Mlp.java
package com.yzumis.ia.mlp;

import com.yzumis.ia.genetic.Reproducible;
import com.yzumis.ia.neuron.Neuron;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Yzumi on 27/12/2017.
 */
public class Mlp implements Reproducible {

    private final List<List<Neuron>> neuronLayers;

    public Mlp(final int numberOfInputs, final List<Integer> neuronsPerLayer) {
        this.neuronLayers = new ArrayList<>();
        for (int i = 0; i < neuronsPerLayer.size(); i++) {
            final List<Neuron> layer = this.initializeNeuronLayer(numberOfInputs, neuronsPerLayer.get(i), i);
            this.neuronLayers.add(layer);
        }
    }

    public Mlp(final List<List<Neuron>> neuronLayers) {
        this.neuronLayers = neuronLayers;
    }

    private List<Neuron> initializeNeuronLayer(final int numberOfInputs, final int neuronsInLayer, final int numberOflayer) {
        final List<Neuron> ret = new ArrayList<>();
        for(int i = 0; i < neuronsInLayer; i++) {
            final Neuron neuron;
            if(numberOflayer == 0) {
                neuron = new Neuron(numberOfInputs);
            } else {
                final int middleLayerNumberOfInputs = this.neuronLayers.get(numberOflayer - 1).size();
                neuron = new Neuron(middleLayerNumberOfInputs);
            }
            ret.add(neuron);
        }
        return ret;
    }

    public List<Double> calculateOutputs(final List<Double> inputs) {
        for(int i = 0; i < this.neuronLayers.size(); i++) {
            for(int j = 0; j < this.neuronLayers.get(i).size(); j++) {
                if( i == 0) {
                    this.neuronLayers.get(i).get(j).calculateOutput(inputs);
                } else {
                    final List<Double> middleLayerInputs = this.getLayerOutputs(this.neuronLayers.get(i - 1));
                    this.neuronLayers.get(i).get(j).calculateOutput(middleLayerInputs);
                }
            }
        }
        return this.getLayerOutputs(this.neuronLayers.get(this.neuronLayers.size() -1));
    }

    private List<Double> getLayerOutputs(final List<Neuron> neuronLayer) {
        final List<Double> ret = new ArrayList<>();
        for(final Neuron neuron : neuronLayer) {
            ret.add(neuron.getOutput());
        }
        return ret;
    }

    @Override
    public Reproducible reproduce(final Reproducible reproducible, final double mutationRate) {
        final List<List<Neuron>> childNeuronLayers = new ArrayList<>();
        for(int i = 0; i < this.neuronLayers.size(); i++) {
            final List<Neuron> childNeuronLayer = new ArrayList<>();
            for(int j = 0; j < this.neuronLayers.get(i).size(); j++) {
                final Neuron neuron = Neuron.reproduce(this.neuronLayers.get(i).get(j), ((Mlp)reproducible).neuronLayers.get(i).get(j), mutationRate);
                childNeuronLayer.add(neuron);
            }
            childNeuronLayers.add(childNeuronLayer);
        }
        return new Mlp(childNeuronLayers);
    }

}
